juju/juju
# Add local resource paths to bundle spec.

## Description of change

Fixes referenced bug

Bundles can contain many charms. Those charms can require many different resources. Currently Juju allows users to specify a local path to those resources using the `juju attach` command. However, for bundles which require the specification of many resources, it can be burdensome and error-prone to attach every resource individually using `juju attach`. This commit adds the ability for users to specify paths to local resources in the bundle specification itself.  

## How do we verify that the change works?

First enter a working directory of your choosing then download a very simple bundle:
```bash
wget -O simple-bundle.zip https://api.jujucharms.com/charmstore/v5/~kwmonroe/simple-bundle/archive
```
Unizp the bundle
```bash
unzip simple-bundle.zip -d simple-bundle
```

We will also need an ubuntu charm:

```
wget -O ubuntu.zip https://api.jujucharms.com/charmstore/v5/ubuntu-10/archive
```
Unizp the charm
```bash
unzip ubuntu.zip -d ubuntu
```

Open the bundle.yaml file for editing to include a resource:
```
nano ./simple-bundle/bundle.yaml
```

Edit the file to match the following content:
```
services:
  ubuntu:
    charm: "<absolute path to local ubuntu charm>"
    num_units: 1
    annotations:
      gui-x: "550"
      gui-y: "400"
    resources:
      software: "./simple-bundle/simple-bundle.zip"
series: xenial
```

Notice how we added the `software: "./simple-bundle/simple-bundle.zip"` under our newly added "resources:" property. This tells `juju` that the bundle has an application whose resource is to be found at the local path  `./simple-bundle/simple-bundle.zip`.

Edit the ubuntu charm to add a `software` resource. The file should look like below:

```
"name": "ubuntu"
"summary": "A pristine Ubuntu Server"
"maintainer": "Charmers <juju@lists.ubuntu.com>"
"description": |
  This simply deploys the Ubuntu Cloud/Server image
"tags":
- "misc"
"series":
- "xenial"
- "trusty"
- "precise"
resources:
  software:
    type: file
    filename: simple-bundle.zip
    description: "One line description that is useful when operators need to push it."
```

Deploy our simple bundle
```
juju deploy simple-bundle
```

You should get the following error:

```
ERROR cannot deploy bundle: file for resource "software": stat ./simple-bundle/simple-bundle.zip: no such file or directory
```

We get the above error because the resource does not exist at the path that we specified in our bundle.yaml file. This shows that `juju` is taking into account the path specified in our bundle.yaml. 

Move the resource to the correct path:
```
mv ./simple-bundle.zip ./simple-bundle
```

Try to deploy the bundle again:
```
juju deploy ./simple-bundle
```

Everything should work fine:

```
Deploying charm "local:xenial/ubuntu-16"
added resource software
Deploy of bundle completed.
```

View the ubuntu applications resources:
```
juju resources ubuntu
```

You should see output similar to the following:

```
[Service]
Resource  Supplied by  Revision
software  admin        2017-27-07T21:15
```

## Does it affect current user workflow? CLI? API?

Yes, users can now use bundle specifications to specify the path of local resources.

## Bug reference

[lp#1623217](https://bugs.launchpad.net/juju/+bug/1623217)
